### DISCUSSION TITLE
Implementing a distance between points using a non-linear function

### DISCUSSION THREAD

**Discussion by maxomous** [google/or-tools#3220] [CP-SAT questions] [answered]
I'm attempting to use Google's OR-Tools to produce a distance constraint between 2x 2D points (p0 & p1).

This becomes the non-linear expression: 
(p1x - p0x)² + (p1y - p0y)² = d²

I have attempted to produce this with the CP-Sat solver using the following code:

```cpp	
	int64_t d = 50;
	int64_t d_sq = d*d;
	
	// p0 at (10, 20)
	IntVar p0x = cp_model.NewConstant(0);
	IntVar p0y = cp_model.NewConstant(0);
	
	// p1 is to be determined by solver
	IntVar p1x = cp_model.NewIntVar({ int64_t_min, int64_t_max });
	IntVar p1y = cp_model.NewIntVar({ int64_t_min, int64_t_max });
	
	// x1-x0 and y1-y0
	IntVar xDif = cp_model.NewIntVar({ -d, d });
	IntVar yDif = cp_model.NewIntVar({ -d, d });
	cp_model.AddEquality(xDif, p1x - p0x);
	cp_model.AddEquality(yDif, p1y - p1y);

	// xdif_sq = xdif^2  &  ydif_sq = ydif^2
	IntVar xDif_sq = cp_model.NewIntVar({ 0, d_sq }); //obvious solution is to add a tolerance range to d. But I'm having an issue 
	IntVar yDif_sq = cp_model.NewIntVar({ 0, d_sq });
	cp_model.AddMultiplicationEquality(xDif_sq, { xDif, xDif });
	cp_model.AddMultiplicationEquality(yDif_sq, { yDif, yDif });

	// dif_sq_sum = xdif^2 + ydif^2
	IntVar Dif_sq_sum = cp_model.NewIntVar({ 0, 2*d_sq });
	cp_model.AddEquality(Dif_sq_sum, xDif_sq + yDif_sq);
	
    // make both sides of equation equal 
	cp_model.AddEquality(iv_Dif_sq_sum, d_sq);
```	
The problem is that because the solver only accepts integers, it can only calculate the position of p1 when the all parts are integer values. 

For example, 3² + 4² = 5² works fine.

But I want to make it work for when a term is non-integer? a² + b² = (almost) c²

The obvious solution is to add a tolerance range to d. But I'm having an issue trying to achieve this as when I do, the status becomes infeasible. Here is the code I've tried.
    
	IntVar iv_Error;
	cp_model.AddMultiplicationEquality(iv_Dif_sq_sum, { iv_Error, iv_Error });
	cp_model.AddLessThan(iv_Error, d + 1);
	cp_model.AddGreaterThan(iv_Error, d - 1);

If anybody has any good suggestions I would be very grateful!
Thanks

**Comment by jmarca** [comment]
Maybe multiply by 1000 and round?

James

**Reply by maxomous** [reply]
I am sort of doing that already, as I am working to 0.01mm accuracy, I am multiplying everything by 100 and then back again. The issue with this approach is that it seems to be excessively slow (~30s) whereas I am aiming to be able to update this in real-time.

**Comment by lperron** [comment]
what do you want to do with the product ? minimize it ?
In that case, you just need c^2 >= a^2 + b^2

**Reply by lperron** [reply]
or you write `(c + 1) ^ 2 > a^2 + b^2` && `c^2 <= a^2 + b^2`

**Reply by maxomous** [reply]
Sorry for the late response, I have been on holiday.

There are potentially many constraints like this between points and so my thought was to sum the errors of each constraint and minimise that, does this sound like a logical approach?

I have gone with your suggestion of (c + 1) ^ 2 > a^2 + b^2 && c^2 <= a^2 + b^2 and the error is currently the difference between the 2 sides rather than the actual difference in distance. This seems to work quite nicely but I am wondering whether I need the negative side also (e.g. c - 1)?

Another issue now however is that it can run quite slow and I think this is because of the domain size. I am attempting to produce a 0.01mm accuracy and to achieve this, I multiply my values by 100 and then back again after (and the domain is scaled accordingly). At this accuracy, it takes about 30s to complete! For 0.1mm accuracy, it takes about 2s. What other ways can I attempt to speed up this process?

Thanks again!

**Comment by lperron** [comment, marked as answer]
Implemented in https://github.com/google/or-tools/blob/stable/examples/python/spread_robots_sat.py
